sympy/algebras/quaternion.py

```diff
@@ -0,0 +1,545 @@
+# References :
+# http://www.euclideanspace.com/maths/algebra/realNormedAlgebra/quaternions/
+# https://en.wikipedia.org/wiki/Quaternion
+
+from __future__ import print_function
+
+from sympy.core.expr import Expr
+from sympy import Rational
+from sympy import re, im, conjugate
+from sympy import sqrt, sin, cos, acos, asin, exp, ln
+from sympy import trigsimp
+from sympy import diff, integrate
+from sympy import Matrix, Add, Mul
+from sympy import symbols, sympify
+from sympy.printing.latex import latex
+from sympy.printing import StrPrinter
+
+
+class Quaternion(Expr):
+    """Provides basic quaternion operations.
+    Quaternion objects can be instantiated as Quaternion(a, b, c, d)
+    as in (a + b*i + c*j + d*k).
+
+    Example
+    ========
+
+    >>> from sympy.algebras.quaternion import Quaternion
+    >>> q = Quaternion(1, 2, 3, 4)
+    >>> q
+    1 + 2*i + 3*j + 4*k
+    """
+    _op_priority = 11.0
+
+    is_commutative = False
+
+    def __new__(cls, a=0, b=0, c=0, d=0, real_elements = True):
+        a = sympify(a)
+        b = sympify(b)
+        c = sympify(c)
+        d = sympify(d)
+
+        if any(i.is_commutative is False for i in [a, b, c, d]):
+            raise ValueError("arguments have to be commutative")
+        else:
+            if any(i.is_real is False for i in [a, b, c, d]):
+                real_elements = False
+                
+            obj = Expr.__new__(cls, a, b, c, d)
+
+            obj._a = a
+            obj._b = b
+            obj._c = c
+            obj._d = d
+            obj._real_elements = real_elements
+            return obj
+            
+
+    @property
+    def a(self):
+        return self._a
+
+    @property
+    def b(self):
+        return self._b
+
+    @property
+    def c(self):
+        return self._c
+
+    @property
+    def d(self):
```

Comment:
Are a, b, c, d standard when working with quaternions? It seems it would get confusing to keep track of which corresponds to the usual i, j, k.

Reply:
The Wikipedia page uses a, b, c, d in the definition.
https://en.wikipedia.org/wiki/Quaternion

Reply:
But are these incidental or commonly used? You have to use something as the scalar variables, but that doesn't necessarily mean the variable names are part of the definition. 

I don't use quaternions myself, so I don't know. Do people immediately associate in their heads something like c <=> j?

Reply:
The variable names are not part of the definition.
The two most commonly used conventions I came across are
1. a + b*i + c*j + d*k
2. w + x*i + y*j + z*k